File: server/services/otp_service.py
import random
import logging
from datetime import datetime, timedelta
from database.otp import save_otp, find_otp, delete_otp, increment_attempts
from services.email_service import send_otp_email

logger = logging.getLogger(__name__)

def generate_otp():
    """Generates a secure 6-digit random number."""
    return str(random.randint(100000, 999999))

def generate_and_send_otp(email):
    """Generates, saves, and emails a 6-digit OTP code to the user."""
    email_clean = email.strip().lower()
    
    # Check cooldown before sending
    can_resend, remaining_seconds = check_resend_cooldown(email_clean)
    if not can_resend:
        return False, f"Please wait {remaining_seconds} seconds before requesting a new OTP."
        
    otp = generate_otp()
    try:
        # Save OTP to DB
        save_otp(email_clean, otp, expires_in_minutes=5)
        # Send Email
        sent = send_otp_email(email_clean, otp)
        if sent:
            return True, "Verification code sent to your email."
        else:
            return False, "Failed to send verification email. Please try again."
    except Exception as e:
        logger.error(f"Error in generate_and_send_otp for {email_clean}: {e}")
        return False, "An unexpected error occurred. Please try again."

def verify_otp_code(email, code):
    """Verifies the code against the database. Tracks attempts and enforces constraints."""
    email_clean = email.strip().lower()
    otp_doc = find_otp(email_clean)
    
    if not otp_doc:
        return False, "OTP has expired or does not exist. Please request a new code."
        
    attempts = otp_doc.get("attempts", 0)
    if attempts >= 5:
        # Delete OTP to force resending
        delete_otp(email_clean)
        return False, "Maximum verification attempts exceeded. Please request a new OTP."
        
    # Increment attempts
    increment_attempts(email_clean)
    
    if otp_doc["otp"] == str(code).strip():
        # Clean up database entry
        delete_otp(email_clean)
        return True, "Email verified successfully."
    else:
        remaining_attempts = 5 - (attempts + 1)
        if remaining_attempts <= 0:
            delete_otp(email_clean)
            return False, "Too many incorrect attempts. This OTP has been invalidated."
        return False, f"Invalid verification code. {remaining_attempts} attempts remaining."

def check_resend_cooldown(email):
    """Checks if the email is allowed to receive a new OTP or is still on cooldown (60 seconds)."""
    email_clean = email.strip().lower()
    otp_doc = find_otp(email_clean)
    
    if not otp_doc:
        return True, 0
        
    created_at = otp_doc.get("created_at")
    if not created_at:
        return True, 0
        
    time_diff = datetime.utcnow() - created_at
    if time_diff < timedelta(seconds=60):
        remaining = 60 - int(time_diff.total_seconds())
        return False, remaining
        
    return True, 0

File: backend/src/services/otpService.js
import { OtpVerification } from "../models/OtpVerification.js";
import { emailService } from "./emailService.js";
import { logger } from "../utils/logger.js";

function generateCode() {
  return String(Math.floor(100000 + Math.random() * 900000));
}

export const otpService = {
  /**
   * Checks if an email is on cooldown (60 seconds).
   */
  async checkResendCooldown(email) {
    const cleanEmail = email.trim().toLowerCase();
    const existing = await OtpVerification.findOne({ email: cleanEmail });

    if (!existing || !existing.created_at) {
      return { canResend: true, remainingSeconds: 0 };
    }

    const elapsedMs = Date.now() - new Date(existing.created_at).getTime();
    if (elapsedMs < 60 * 1000) {
      const remainingSeconds = Math.ceil((60 * 1000 - elapsedMs) / 1000);
      return { canResend: false, remainingSeconds };
    }

    return { canResend: true, remainingSeconds: 0 };
  },

  /**
   * Generates, stores, and sends OTP.
   */
  async generateAndSendOtp(email) {
    const cleanEmail = email.trim().toLowerCase();
    const { canResend, remainingSeconds } = await this.checkResendCooldown(cleanEmail);

    if (!canResend) {
      return {
        success: false,
        message: `Please wait ${remainingSeconds} seconds before requesting a new OTP.`,
      };
    }

    const otp = generateCode();
    const expiresAt = new Date(Date.now() + 5 * 60 * 1000); // 5 minutes

    try {
      // Upsert OTP record
      await OtpVerification.deleteMany({ email: cleanEmail });
      await OtpVerification.create({
        email: cleanEmail,
        otp,
        attempts: 0,
        expires_at: expiresAt,
      });

      const sent = await emailService.sendOtpEmail(cleanEmail, otp);
      if (sent) {
        return { success: true, message: "Verification code sent to your email." };
      }
      return { success: false, message: "Failed to send verification email. Please try again." };
    } catch (err) {
      logger.error(`Error in generateAndSendOtp for ${cleanEmail}: ${err.message}`);
      return { success: false, message: "An unexpected error occurred. Please try again." };
    }
  },

  /**
   * Verifies the OTP code.
   */
  async verifyOtpCode(email, code) {
    const cleanEmail = email.trim().toLowerCase();
    const cleanCode = String(code).trim();

    const record = await OtpVerification.findOne({
      email: cleanEmail,
      expires_at: { $gt: new Date() },
    });

    if (!record) {
      return {
        success: false,
        message: "OTP has expired or does not exist. Please request a new code.",
      };
    }

    if (record.attempts >= 5) {
      await OtpVerification.deleteMany({ email: cleanEmail });
      return {
        success: false,
        message: "Maximum verification attempts exceeded. Please request a new OTP.",
      };
    }

    // Increment attempts
    record.attempts += 1;
    await record.save();

    if (record.otp === cleanCode) {
      await OtpVerification.deleteMany({ email: cleanEmail });
      return { success: true, message: "Email verified successfully." };
    } else {
      const remaining = 5 - record.attempts;
      if (remaining <= 0) {
        await OtpVerification.deleteMany({ email: cleanEmail });
        return {
          success: false,
          message: "Too many incorrect attempts. This OTP has been invalidated.",
        };
      }
      return {
        success: false,
        message: `Invalid verification code. ${remaining} attempts remaining.`,
      };
    }
  },
};
